### DISCUSSION TITLE
Eldoc: use eldoc-echo-area-use-multiline-p instead of :echo as in 1.14

### DISCUSSION THREAD

**Discussion by yalfg** [joaotavora/eglot#1252] [Ideas]
Hi,

Eglot 1.14 added support for the Eldoc `:echo` feature. The result has been to truncate Eldoc feedback in the modeline to a single line (at least for me, but from the code it seems normal).

On a tall screen I actually like the multi-lines output, but I haven't seen an easy way to change back to the previous behavior. The `:echo` output simply takes priority, and even telling Eldoc to use multiple lines in the echo area has no effect.

Removing the `:echo` addition and using `eldoc-echo-area-use-multiline-p` set to 1 would achieve the same result, but would leave the choice to the user. By putting this in the "stay out of" list, a user could easily have multiple lines in the echo (or whatever ones favorite Eldoc configuration is) while still have the same behavior as 1.14 by default.

The documentation in Eldoc is not super clear to me, but with the existence of `eldoc-echo-area-use-multiline-p` any echo area specific content using `:echo` should strive to have more compact content, but not in reducing the number of shown lines: this is already covered by the user controllable `eldoc-echo-area-use-multiline-p`.

WDYT? Or is there any other way I missed to both get eglot LSP feedback in Eldoc, and the echo area user "heigth control"?

Thanks for you time!

**Comment by joaotavora** [comment]
You're not missing anything.  The change you're referring to does remove an existing behavior which is now harder to get back.  All the feedback I had to the previous behavior was overwhelmingly in the sense that it mostly wasn't useful in the vast majority of situations.  But I suspected someone somewhere actually enjoyed it in some specific situation, and you're here to prove that.

Personally, I think it makes much more sense like this.  Even a multiline echo area is (with `eldoc-echo-area-use-multiline-p`) is, by design, supposed to be a size-limited outlet for at-point documentation information. 

But the echo area is not the only display outlet for ElDoc.  You can create however many you want, just look at `eldoc-display-functions`.

But even without doing that, have you tried keeping the `*eldoc*` buffer showing in some window?  This buffer is another outlet which is by nature not so limiting.  I find myself doing `C-h .` when the echo are doesn't show everything I need, and then keeping this automatically-updating `*eldoc*` buffer visible in some window as I move around.

**Reply by yalfg** [reply]
Yes, I tried the dedicated window and it's not as convenient for my use case. I think it's a small screen vs. big/multi screens thing.

For a laptop screen, with precious vertical space, it makes sense to have a terse echo area output. And this is probably the most common (and conservative) case, so it makes sense to me as a default.

But when I noticed the output shortening was on a dual HiDPI screen configuration. Vertical space is plenty, and I have several frames for different contexts. Having a fixed eldoc window is not so convenient then: it's one more window to manage and can be far from the active frame. Whereas a "rich" echo area output follows the active frame, is not a big deal even if not so relevant and disappear when useless. So it's not intrusive due to the vertical space, and all automatic. Even if it's not the most common case, this kind of set-up is still relatively common I guess.

No fixed configuration can match all needs for sure, but it's nice to be able to easily adapt to one's case. The issue here is that the use of `:echo` filters at the source, in the backend. It's not so easy to get back to the previous case: the only thing that come to my mind is to advice the two eglot eldoc generation functions to strip `:echo` and associated content. Not impossible, but a high bar for a lot of users I guess (not trivial to me, but I think manageable ;).

This is why I thought about `eldoc-echo-area-use-multiline-p`: set to 1, it achieves the same result but then it's easy for a user to change the behavior.

There's also the principle of least surprise (at least for me!): when I noticed the change, I looked around and found `eldoc-echo-area-use-multiline-p`. I set it to multiple lines and was surprised when nothing changed. Well I dug deeper and learnt more, and now I know I shouldn't have expected different (the default allows for multiple lines).

This is more a comment on eldoc, but both `:echo` and `eldoc-echo-area-use-multiline-p` both target the same echo area and can step on each other toes. It seems more flexible to leave the total "height control" to the easy to change multi-lines variable if possible, and use `:echo` for other ways to control terseness? Just my opinion...

Thanks for your time.

**Reply by joaotavora** [reply]
> This is more a comment on eldoc, but both :echo and eldoc-echo-area-use-multiline-p both target the same echo area and can step on each other toes. It seems more flexible to leave the total "height control" to the easy to change multi-lines variable if possible, and use :echo for other ways to control terseness? Just my opinion...

The two things are quite different. `:echo` is an implementor's interface for a specific documentation function (of which there can be many) and `eldoc-e-a-u-m-p` is a user interface for the echo area as a whole when used by ElDoc. So both have their place.

You are in effect asking for a Eglot option meaning "make eglot documentation functions spit out verbose stuff to echo area". It could make sense and wouldn't be very hard to implement. But I haven't seen much demand for it -- you're the first to ask for it -- and I've seen lots of demand against it. So I'd rather wait and see. In the meantime you could implement your own `eldoc-display-function` that ignores the `:echo` request.

**Reply by yalfg** [reply]
Ok, filtering `:echo` and its parameter not at the source but at eldoc display level. I'll try this, thanks!

**Comment by yalfg** [comment]
For anyone interested, the following advice Eldoc echo area display function to strip `:echo` and restore the previous behavior. It's not specific to elgot and a bit brute force, but should be easy to adapt to anyone taste:

```elisp
(defun my/strip-echo (args)
  "Filter :echo from the eldoc plist in ARGS.
This is used as an filtering advice to `eldoc-display-in-echo-area', and disable the new :echo feature."
  (let* ((docs (car args))
         (interact (cadr args))
         (newdocs (mapcar
                   (lambda (doc)
                     (when doc
                       (let ((string (car doc))
                             (plist (cdr doc)))
                         (cl-remf plist :echo)
                         (cons string plist))))
                   docs)))
    (list newdocs interact)))
(advice-add #'eldoc-display-in-echo-area :filter-args #'my/strip-echo)
```

Thanks
